draft-irtf-cfrg-frost.md

```diff
@@ -214,22 +214,45 @@ FROST requires the use of a cryptographically secure hash function, generically
 written as H, which functions effectively as a random oracle. For concrete
 recommendations on hash functions which SHOULD be used in practice, see
 {{ciphersuites}}. Using H, we introduce three separate domain-separated hashes,
-H1, H2, and H3, where H1 and H2 map arbitrary byte strings to Scalar elements of
+H1, H2, H3, and H4, where H1, H2, and H4 map arbitrary byte strings to Scalar elements of
 the prime-order group scalar field, and H3 is an alias for H with a domain separator. 
-The details of H1, H2, and H3 vary based on ciphersuite. See {{ciphersuites}}
+The details of H1, H2, H3, and H4 vary based on ciphersuite. See {{ciphersuites}}
 for more details about each.
 
 # Helper functions {#helpers}
 
 Beyond the core dependencies, the protocol in this document depends on the
 following helper operations:
 
+- Nonce generation, {{dep-nonces}}
 - Schnorr signatures, {{dep-schnorr}};
 - Polynomial operations, {{dep-polynomial}};
 - Encoding operations, {{dep-encoding}};
 - Signature binding {{dep-binding-factor}}, group commitment {{dep-group-commit}}, and challenge computation {{dep-sig-challenge}}
 
-This sections describes these operations in more detail.
+These sections describes these operations in more detail.
+
+## Nonce generation {#dep-nonces}
+
+To hedge against a bad RNG, we generate nocnes by sourcing fresh randomness and 
+combine with the secret key, to create a domain-separated hash function from 
+the ciphersuite hash function `H`, `H4`:
+
+~~~
+  nonce_generate(secret):
+  
+  Inputs:
+  - secret, a Scalar in GF(p).
+
+  Outputs: Scalar nonce
+
+  def nonce_generate(secret):
```

Comment:
secret_share? or sk_i, to use notation that we employ elsewhere.

Reply:
If we also want to use this in `schnorr_signature_generate()` which I do in this PR, it has to be either a singleton full secret key or a threshold secret key share